gh-111964: Add _PyRWMutex a "readers-writer" lock

colesbury opened the pull request:
This adds `_PyRWMutex`, a "readers-writer" lock, which wil be used to
serialize global stop-the-world pauses with per-interpreter pauses.


<!-- gh-issue-number: gh-111964 -->
* Issue: gh-111964
<!-- /gh-issue-number -->

colesbury committed:
gh-111964: Add `_PyRWMutex` a "readers-writer" lock

This adds `_PyRWMutex`, a "readers-writer" lock, which wil be used to
serialize global stop-the-world pauses with per-interpreter pauses.

colesbury committed:
Add comments and ASCII diagram based on review.

Also, fix bug where `_PyRWMutex_Lock()` swapped the
return/continue conditions leading to an extra iteration of the loop.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -213,6 +213,41 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+// The design is optimized for simplicity of the implementation. The lock is
+// not fair: if fairness is desired, use an additional PyMutex to serialize
+// writers. The lock is also not reentrant.
+typedef struct {
+    uintptr_t bits;
+} _PyRWMutex;
```
Following up on https://github.com/python/cpython/pull/112471#discussion_r1418158478:

> > If I've understood correctly, this is essentially a Py_Mutex, with the extra bits used for a counter.
> 
> Yes, in the sense that both use two bits for state (locked and has_parked) and those bits are in the same positions. I think it's more helpful to think of it as "similar to a PyMutex" than embedding a PyMutex.

Ah, that helps.  I was definitely thinking of `_PyRWMutex` as, effectively, a subtype of `PyMutex`.

> I would write the representation as:
> 
> ```c
> // e.g., 62 on 64-bit platforms
> #define READER_COUNT_BITS (sizeof(uintptr_t)*CHAR_BIT - 2)
> 
> struct {
>   bool is_write_locked: 1;
>   bool has_parked: 1;
>   uintptr_t reader_count: READER_COUNT_BITS;
> };
> ```
> 
> > Ideally, _PyRWMUTEX_MUTEX_BITS would be the bits needed for the largest bit flag used for PyMutex. However, in this case we're using a subset of the possible PyMutex flags...
> 
> I'm not sure I understand what you mean. Ignoring the reader count for a moment, both `PyMutex` and `_PyRWMutex` only use two bits for state: is_locked? and has_parked?.

<details>
<summary>(expand for a now irrelevant explanation)</summary>

----

I suppose I was thinking in terms of Liskov substitution, i.e. casting a `_PyRWMutex` to `PyMutex` and passing it to `PyMutex`-specific functions.

Based on the `PyMutex` definition:

https://github.com/python/cpython/blob/480b4b359d710c74f31e2ffd7fc51750dcec7012/Include/internal/pycore_lock.h#L40-L49

It is effectively:

```C
struct _PyMutex {
  bool is_write_locked: 1;
  bool has_parked: 1;
  bool once_initialized: 1;
  uint8_t _padding: 5;
};
```

The equivalent for `_PyRWMutex` only has the first two fields, so we could use any `PyMutex`-specific functions for `_PyRWMutex,, but only those that are limited to those two fields.  Naturally it isn't obvious which functions match that criteria without reading the code.  It would be better to take advantage of the C compiler for the distinction.  That's assuming there are functions we could use for both types.

----

</details>

All that said, it is clear to me now that, while `PyMutex` and `_PyRWMutex` do share an underlying design[^1], they don't really share any code. [2]  They only happen to share the layout of their first two bits, which is strictly because we're using the same two bit flag macros (`_Py_LOCKED` and `_Py_HAS_PARKED`) for both.  Is that all correct?

Assuming there isn't any potential for sharing code between them, I think we're good to go on this.  Perhaps we could avoid sharing the bit flag macros. [^3]  That was a main source of my original confusion.

[^1]: atomic int ops + flag bits + order of operations to preserve correctness
[^2]: They do both call `_PyParkingLot_Park()`, but that isn't relevant since the parking lot doesn't care about the layout of the memory at the given address.
[^3]: e.g. `_Py_LOCKED` -> `_PyMutex_LOCKED` and `_PyRWMutex_LOCKED`

ericsnowcurrently replied:
I see you already added `_Py_WRITE_LOCKED` to help avoid the appearance of coupling.  That definitely helps.  Using a type-specific prefix might be even clearer and take care of _Py_HAS_PARKED too:

* `_Py_LOCKED` -> `_PyMutex_LOCKED`
* `_Py_WRITE_LOCKED` -> `_PyRWMutex_WRITE_LOCKED`
* `_Py_HAS_PARKED` -> `_PyMutex_HAS_PARKED`
* `_Py_HAS_PARKED` -> `_PyRWMutex_HAS_PARKED` (or even `_PyRWMutex_WRITE_HAS_PARKED` or `_PyRWMutex_HAS_PARKED_WRITER`)

colesbury replied:
I would prefer to avoid the type prefixes because they add a lot of verbosity, which makes it harder to see the important parts of the algorithm. They'll force a lot of extra line wrapping for otherwise simple operations.

I added `_Py_WRITE_LOCKED` because  "locked" alone is ambiguous in this context -- that bit specifically means write-locked (as opposed to read-locked).

When MSVC finally makes C11 atomics non-experimental, we can probably use a bitfield for `_PyRWMutex` (and most other structs here), which I think will make things both more clear and concise.

ericsnowcurrently replied:
I still think the prefixed names would be clearer, but I get what you're saying and I'm fine with the way things are.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,107 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+{
+    uintptr_t old_bits = _Py_atomic_exchange_uintptr(&rwmutex->bits, 0);
+
+    assert((old_bits & _Py_WRITE_LOCKED) && "lock was not write-locked");
+    assert(rwmutex_reader_count(old_bits) == 0 && "lock was read-locked");
```
perfect

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,107 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+                continue;
+            }
+            return;
+        }
```
Thanks for splitting this up and adding comments.  This is much easier to follow now.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,107 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+_PyRWMutex_RUnlock(_PyRWMutex *rwmutex)
+{
+    uintptr_t bits = _Py_atomic_add_uintptr(&rwmutex->bits, -(1 << _PyRWMutex_READER_SHIFT));
+    bits -= (1 << _PyRWMutex_READER_SHIFT);
```
A mis-applied `_PyRWMutex_RUnlock()` could cause a pseudo-underflow, right?  How about adding an assert here like you did in `_PyRWMutex_Unlock()`?

```suggestion
    uintptr_t bits = _Py_atomic_add_uintptr(&rwmutex->bits, -(1 << _PyRWMutex_READER_SHIFT));
    assert(rwmutex_reader_count(bits) > 0);
    bits -= (1 << _PyRWMutex_READER_SHIFT);
```

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,107 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+            // The lock is write-locked.
+            bits = rwmutex_set_parked_and_wait(rwmutex, bits);
+            continue;
+        }
```
just to make the single-writer/many-readers characteristic extra clear:

```suggestion
        if ((bits & _Py_WRITE_LOCKED)) {
            // A writer already holds the lock.
            bits = rwmutex_set_parked_and_wait(rwmutex, bits);
            continue;
        }
```

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -353,3 +353,107 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
+            // ourselves and wait for the writer to eventually wake us up.
+            bits = rwmutex_set_parked_and_wait(rwmutex, bits);
+            continue;
+        }
```
Can this branch ever be taken?  If a writer has the lock then the earlier branch is taken.  `_Py_HAS_PARKED` is cleared at the beginning of `_PyRWMutex_Unlock()` before `_PyParkingLot_UnparkAll()` gets called, so previously parked readers won't hit this either.  Am I missing something?

colesbury replied:
This branch is called when readers hold the lock and there is a waiting writer. Although `_Py_HAS_PARKED` in general can indicate waiting readers or writers, if the lock is read-locked it must imply a waiting writer because otherwise readers don't block other readers.

ericsnowcurrently replied:
Ah, I get it now.  Thanks for explaining.

colesbury replied:
There's actually one other case that this branch handles that I forgot about: There's a transient state after the last reader unlocks the mutex but before a writer grabs it in which `_Py_HAS_PARKED` is set, but the lock is neither read-locked nor write-locked. It's a transient state because there's guaranteed to be at least one previously waiting writer actively trying to write-lock the mutex.

The behavior is the same in this state -- we still want readers to wait for the writer -- but it means the assertion I added did not hold.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -213,6 +213,41 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
 
+// A readers-writer (RW) lock. The lock supports multiple concurrent readers or
+// a single writer. The lock is write-preferring: if a writer is waiting, then
+// new readers will be blocked. This avoids starvation of writers.
```
How far do we want to go with prioritizing waiting writers?

IIUC, currently there's a race with waiting readers and writers once they are all woken up by `_PyParkingLot_UnparkAll()`.  To see what I mean, let's step through the interaction between a handful of threads sharing a `_PyRWMutex`:

1. writer A takes the lock
2. reader B is parked
3. writer C is parked
4. reader D is parked
5. writer A releases the lock
6. the parked threads wake up and take the next iteration in their respective loops (in `_PyRWMutex_RLock()` or `_PyRWMutex_Lock()`)

Since we clear `_Py_HAS_PARKED` at the beginning of `_PyRWMutex_Unlock()`, *before* the `_PyParkingLot_UnparkAll()` call, any one of the following could happen:

<details>
<summary>(expand)</summary>

* both readers beat the writer
   7. readers B and D take the lock
   8. writer C is parked again
   9. readers B and D release the lock
   10. writer C takes the lock
   11. writer C releases the lock
* one reader beats the writer
   7. reader B or D takes the lock
   8. writer C is parked again
   9. the other reader takes the lock too
   10. readers B and D release the lock
   11. writer C takes the lock
   12. writer C releases the lock
* the writer beats the readers
   7. writer C takes the lock
   8. readers B and D are parked again
   9. writer C releases the lock
   10. readers B and D take the lock
   11. readers B and D release the lock

</details>

Thus, in 2 of 3 cases we fail to prioritize a waiting writer over a waiting reader.

### Possible Solutions

I expect the situation would be slightly better if we were to set `_Py_HAS_PARKED` right after the `rwmutex_set_parked_and_wait()` call in `_PyRWMutex_Lock()`.  There would still be a small race with readers, but there's a better chance a waiting writer gets prioritized.

Another idea is to somehow put the readers in a different parking lot (or bucket) than the writers and unpark waiting writers first.

<details>
<summary>One more idea:</summary>

Perhaps there's a way we can be thorough about prioritizing waiting writers.  Let's revisit the steps we took above, re-using steps 1-6.

Here's how it would look if we cleared `_Py_HAS_PARKED` later than we do now: at the point where we don't have any waiting writers left.  That would looks like one of the following (depending on which thread wins:

* both readers beat the writer
   7. readers B and D are parked again (the `_Py_HAS_PARKED` branch)
   8. writer C takes the lock
   9. writer C releases the lock
   10. `_Py_HAS_PARKED` is cleared
   11. readers B and D take the lock
* one reader beats the writer
   7. reader B or D is parked again (the `_Py_HAS_PARKED` branch)
   8. writer C takes the lock
   9. the other reader is parked again
   10. writer C releases the lock
   11. `_Py_HAS_PARKED` is cleared
   12. readers B and D take the lock
* the writer beats the readers
   7. writer C takes the lock
   8. readers B and D are parked again
   9. writer C releases the lock
   10. `_Py_HAS_PARKED` is cleared
   11. readers B and D take the lock

Step 10 (or 11 in the second scenario) is the one that facilitates always prioritizing waiting writers.  To do so, we'd clear `_Py_HAS_PARKED` later than we do now and only if there were no other writers trying to take the lock:

1. also in `_PyRWMutex_Unlock()` but at the end, after the `_PyParkingLot_UnparkAll()` call, and only if there are no other writers trying to take the lock
2. in `_Py_RWMutex_Lock()` in the success case right before returning

In either case, it would require keeping a count of writers, like we do for readers (and checking that count in this branch).

</details>

colesbury replied:
I don't think we want to go any further with prioritizing writers. The goal was narrow: to avoid writer starvation when there are a lot of readers. If we were to further optimize this, we'd probably want to do the opposite: handoff to waiting readers in `_PyRWMutex_Unlock()` so they get a turn too (that's what Go's sync.RWMutex does, for example).

I don't think this is worth optimizing further. For some context, I didn't use this at all in the nogil forks and I'm not sure there will be other uses of this API. That's one of the reasons for "optimizing for simplicity of the implementation".

ericsnowcurrently reviewed (changes requested):
Thanks for taking the time to split this out and address the other review comments I left on gh-112471.  I've left a few here, including what might be a bug.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -213,6 +213,41 @@ _PyOnceFlag_CallOnce(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
 
+// A readers-writer (RW) lock. The lock supports multiple concurrent readers or
+// a single writer. The lock is write-preferring: if a writer is waiting, then
+// new readers will be blocked. This avoids starvation of writers.
```
It might be nice to also briefly note that another name for this is a "shared mutex", with distinct "shared" vs. "exclusive" operations.  (Basically, what you said in https://github.com/python/cpython/pull/112471#discussion_r1418183599.)

colesbury replied:
I added a comment.

colesbury committed:
Update comments and add assertions.

Also describe _PyRWMutex using bit patterns instead of ASCII art box. This
matches how PyMutex is described.

ericsnowcurrently reviewed (approved):
LGTM

colesbury committed:
Fix typo

colesbury committed:
Remove failing assertion

colesbury committed:
Add comment describing state with _Py_HAS_PARKED set
